src/hotspot/share/runtime/reflection.cpp

```diff
@@ -767,8 +767,8 @@ static Handle new_type(Symbol* signature, Klass* k, TRAPS) {
 
 oop Reflection::new_method(const methodHandle& method, bool for_constant_pool_access, TRAPS) {
   // Allow sun.reflect.ConstantPool to refer to <clinit> methods as java.lang.reflect.Methods.
-  assert(!method()->is_initializer() ||
-         (for_constant_pool_access && method()->is_static()),
+  assert(!method()->is_object_initializer() &&
+         (for_constant_pool_access || !method()->is_static_initializer()),
          "should call new_constructor instead");
```

Comment:
Nit: existing -The assert message isn't really correct

Reply:
Yeah, it is a bit odd. I thought to leave the messages alone, but we can massage them as well. Should be done in new commit.